config/v1/types_ingress.go

```diff
@@ -91,7 +91,7 @@ type IngressSpec struct {
 type ConsumingUser string
 
 // Hostname is an alias for hostname string validation.
-// +kubebuilder:validation:Format=hostname
+// +kubebuilder:validation:Pattern="^([a-zA-Z0-9\\p{S}\\p{L}]((-?[a-zA-Z0-9\\p{S}\\p{L}]{0,62})?)|([a-zA-Z0-9\\p{S}\\p{L}](([a-zA-Z0-9-\\p{S}\\p{L}]{0,61}[a-zA-Z0-9\\p{S}\\p{L}])?)(\\.)){1,}([a-zA-Z0-9-\\p{L}]){2,63})$"
```

Comment:
A few suggestions and observations:

* You can use ``` `` ``` to avoid the need to escape backslashes:

```suggestion
// +kubebuilder:validation:Pattern=`^([a-zA-Z0-9\p{S}\p{L}]((-?[a-zA-Z0-9\p{S}\p{L}]{0,62})?)|([a-zA-Z0-9\p{S}\p{L}](([a-zA-Z0-9-\p{S}\p{L}]{0,61}[a-zA-Z0-9\p{S}\p{L}])?)(\.)){1,}([a-zA-Z0-9-\p{L}]){2,63})$`
```

* I don't believe we should allow a leading or trailing digit or dash in the TLD:

```suggestion
// +kubebuilder:validation:Pattern=`^([a-zA-Z0-9\p{S}\p{L}]((-?[a-zA-Z0-9\p{S}\p{L}]{0,62})?)|([a-zA-Z0-9\p{S}\p{L}](([a-zA-Z0-9-\p{S}\p{L}]{0,61}[a-zA-Z0-9\p{S}\p{L}])?)(\.)){1,}[a-zA-Z\p{L}][a-zA-Z0-9-\p{L}]{0,61}[a-zA-Z\p{L}])$`
```

* I think [the upstream pattern](https://github.com/kubernetes/kubernetes/blob/9f791bbf5b0e27521a33ab63f6a4c16a677e20f4/vendor/k8s.io/kube-openapi/pkg/validation/strfmt/default.go#L58) is incorrect because it does not allow a dash in an unqualified TLD; for example, `foo-bar.tld` is accepted, but `foo-bar` is rejected.  Would the API and router admit `foo-bar`?  If so, we should allow it here too, something like this (this is getting harried and needs heaps of scrutiny):  

```suggestion
// +kubebuilder:validation:Pattern=`^([a-zA-Z0-9\p{S}\p{L}][a-zA-Z0-9-\p{S}\p{L}]{0,61}[a-zA-Z\p{S}\p{L}]|([a-zA-Z0-9\p{S}\p{L}](([a-zA-Z0-9-\p{S}\p{L}]{0,61}[a-zA-Z0-9\p{S}\p{L}])?)(\.)){1,}[a-zA-Z\p{L}][a-zA-Z0-9-\p{L}]{0,61}[a-zA-Z\p{L}])$`
```

* The [Route API validation](https://github.com/openshift/openshift-apiserver/blob/6ae368b1f0170d8b96dcc926fe13a541ba6e3bd1/pkg/route/apis/route/validation/validation.go#L65-L69) does not allow upper-case, so I'm not so sure that including the `A-Z` character class is appropriate.  Punycode should be allowed for IDN, but allowing digits and dashes should suffice for that; I'm not sure we should be using the `\p{S}`, and `\p{L}` character classes here.  We generally cannot tighten validation of a shipped API, but maybe we can in this case—if a user has an installed cluster with a host name that requires the `A-Z`, `\p{S}`, or `\p{L}` character class for the pattern to match, do the API's and router's validation render the cluster broken anyway?  

* I'm beginning to wonder whether it would be better to loosen the validation here to simplify the validation at the cost of potentially allowing users to configure invalid names; the API or router would still ultimately reject the invalid names, so maybe it would suffice to do some basic validation here, and if users give us garbage that's hard to verify is garbage, those users will just have do to a little more legwork to figure out why the garbage input weren't resulting in a working route.

Reply:
I agree, the upsteam pattern has a number of issues I did not address in this PR.  I tried to minimize changes in fixing the TLD issue, for this shipped API.  I can loosen the validation.

There is no `[Route](https://github.com/openshift/api/blob/master/route/v1/types.go#L40)`-style validation on `[componentRoutes](https://github.com/openshift/api/blob/be5cc87d5a14444bfe6e6bdc813519d7c73acad5/config/v1/types_ingress.go#L105)` - they are different types.  [This PR](https://github.com/openshift/cluster-ingress-operator/pull/552/files) introduces a new controller for `ComponentRoutes` and only validates RBAC values.

Reply:
There was some discussion in https://github.com/openshift/enhancements/pull/577#discussion_r560012605